utils/terraform_client.py

```diff
@@ -23,81 +26,144 @@ def __init__(self, body, integration, integration_version):
 
 class TerraformClient(object):
     def __init__(self, integration, integration_version,
-                 integration_prefix, working_dirs):
+                 integration_prefix, working_dirs, thread_pool_size):
         self.integration = integration
         self.integration_version = integration_version
         self.integration_prefix = integration_prefix
         self.working_dirs = working_dirs
+        self.thread_pool_size = thread_pool_size
+        self._log_lock = Lock()
+        init_specs = self.init_init_specs(working_dirs)
+
+        pool = ThreadPool(self.thread_pool_size)
+        results = pool.map(self.terraform_init, init_specs)
+
         tfs = {}
-        for name, wd in working_dirs.items():
-            tf = Terraform(working_dir=wd)
-            return_code, stdout, stderr = tf.init()
-            error = self.check_output(name, return_code, stdout, stderr)
-            if error:
-                return None
+        for name, tf in results:
             tfs[name] = tf
         self.tfs = tfs
 
+    class initSpec(object):
+        def __init__(self, name, wd):
+            self.name = name
+            self.wd = wd
+
+    def init_init_specs(self, working_dirs):
+        init_specs = []
+        for name, wd in working_dirs.items():
+            init_spec = self.initSpec(name, wd)
+            init_specs.append(init_spec)
+        return init_specs
+
+    def terraform_init(self, init_spec):
+        name = init_spec.name
+        wd = init_spec.wd
+        tf = Terraform(working_dir=wd)
+        return_code, stdout, stderr = tf.init()
+        error = self.check_output(name, return_code, stdout, stderr)
+        if error:
+            return name, None
+        return name, tf
+
+    # terraform plan
     def plan(self, enable_deletion):
         errors = False
         deletions_detected = False
-        for name, tf in self.tfs.items():
-            return_code, stdout, stderr = tf.plan(detailed_exitcode=False)
-            error = self.check_output(name, return_code, stdout, stderr)
+
+        pool = ThreadPool(self.thread_pool_size)
```

Comment:
we could probably define the pool as an instance variable and reuse it everytime: `self.pool`

Reply:
done